### DISCUSSION TITLE
Details

### DISCUSSION THREAD

**Discussion by jrbarnes9** [ONSdigital/design-system#148] [Feedback]
Use this place to discuss the [details component](https://ons-design-system.netlify.com/components/details) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by bameyrick** [comment]
It isn't possible to apply custom classes to the details element.

**Reply by jrbarnes9** [reply]
#189 resolved

**Comment by bameyrick** [comment]
@will-madebymoon and I think we should remove the arrow animation as it doesn't add any value.

**Comment by bameyrick** [comment]
When a user closes the details component using the inner button, the summary element should be refocused for screen readers to be able to easily move on to the next item on the page.

**Reply by jrbarnes9** [reply]
#250 resolved

**Comment by bameyrick** [comment]
Buttons used in details components are type submit. These should be type button.

**Reply by jrbarnes9** [reply]
#363 resolved

**Comment by jrbarnes9** [comment]
@will-madebymoon Paragraph content in an expanded details component is getting a bit lost when it's displayed inline with a lot of paragraphs in body content...
<img width="637" alt="Screenshot 2019-06-12 at 08 30 07" src="https://user-images.githubusercontent.com/43346934/59331627-4fc51f80-8cec-11e9-99b8-61f7426dc8cd.png">

Could we consider indenting it or something like e.g....
<img width="645" alt="Screenshot 2019-06-12 at 08 32 15" src="https://user-images.githubusercontent.com/43346934/59331826-c19d6900-8cec-11e9-9db3-f912226e8b28.png">

**Reply by jrbarnes9** [reply]
#580 resolved

**Comment by jrbarnes9** [comment]
Details needs a default bottom margin

**Reply by jrbarnes9** [reply]
Issue #1678 raised

**Comment by bameyrick** [comment]
Maybe we should address all all of these issues when we look at spacing: https://trello.com/c/h0RbO4UV/99-spacing

**Comment by bameyrick** [comment]
Details element cannot be accessed with dragon naturally speaking as a button or as a link

**Reply by jrbarnes9** [reply]
#445 resolved

**Comment by jrbarnes9** [comment]
Details chevron hidden when screen colours inverted (using v12 ZoomFusion)

**Reply by jrbarnes9** [reply]
#453 resolved

**Comment by bameyrick** [comment]
Pointer events need to be turned off on all elements within the summary as GA fails to pick up click events

**Reply by jrbarnes9** [reply]
#508 resolved

**Comment by bameyrick** [comment]
When running in non-js mode, the summary element still renders visually as a link. This should only look like a link when it's running in JS mode.

![Screenshot 2019-07-18 at 10 39 40](https://user-images.githubusercontent.com/8447446/61449288-80802080-a94c-11e9-91a7-df2032a9a601.png)

**Reply by jrbarnes9** [reply]
#556 resolved

**Comment by jrbarnes9** [comment]
`details__title` should use a semantically appropriate heading tag (not a `<span>`) when no js

**Reply by jrbarnes9** [reply]
#615 resolved

**Comment by jrbarnes9** [comment]
Safari rendering issue: font-size rem units ignored on elements following `<details>` and `<summary>` (https://colloq.io/blog/safaris-detailssummary-rem-font-size-issue)

**Reply by jrbarnes9** [reply]
#741 resolved

**Comment by jrbarnes9** [comment]
Collapsible chevron doesn't left align with other page elements when collapsed

<img width="582" alt="Screenshot 2020-09-16 at 11 59 56" src="https://user-images.githubusercontent.com/43346934/93328932-55c17e80-f814-11ea-8d4e-a9c394cf1db8.png">

**Reply by jrbarnes9** [reply]
#990 resolved

**Comment by jrbarnes9** [comment]
Collapsible component no javascript fallback not working 

(Contents should display with a semantic heading above (no chevron or link style) See: #711

**Reply by jrbarnes9** [reply]
#1002 resolved

**Comment by jrbarnes9** [comment]
Collapsible title not rendered at correct font-size when `h2` `titleTag` is applied. Heading tag should only be applied when js is off.
![Screenshot 2020-12-11 at 15 54 50](https://user-images.githubusercontent.com/43346934/101926286-e10b9f80-3bca-11eb-8277-b16812e35a4c.png)

**Reply by jrbarnes9** [reply]
#1238 resolved

**Comment by mr-nick17** [comment]
Could this component use the semantic [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) markup? The browser can hide/show content without JS. The only possible drawback is IE11, the content is still rendered but could potentially lose its semantic meaning

**Reply by jrbarnes9** [reply]
We did use to, but we had to pull it due to a bug (#741) found during Census rehearsal

**Reply by jrbarnes9** [reply]
Added a card to the backlog to document why we don't currently use `<details>`

**Reply by jrbarnes9** [reply]
Agreed to reinstate the `<details>` element to resolve #1996

**Comment by jrbarnes9** [comment]
We should change collapsible `titleTag` default to `h2` because it should always have a semantic heading for no js.

However, the heading tag should only be used when js is off.

Also, when the default `<div>` is used the icon is not vertically centre-aligned...

Collapsible with `<h2>`:
![Screenshot 2021-12-23 at 11 45 05](https://user-images.githubusercontent.com/43346934/147238925-9056a1aa-263c-4803-a419-eb0d9c2cf62e.png)

Collapsible with `<div>`:
![Screenshot 2021-12-23 at 11 44 33](https://user-images.githubusercontent.com/43346934/147238952-3f7dacf9-a7e8-42ba-9dd9-9a1bdda93f1c.png)

**Reply by jrbarnes9** [reply]
#1808 fixed

**Comment by James-Nitron** [comment]
A recent DAC audit has flagged a level A issue with the collapsible component. 

WCAG Reference: 
4.1.1 Parsing (Level A)

The ‘Show’ elements are not nested accordingly to the HTML5 standards. The ‘Show’ 
elements are buttons that are nested within a link, and this is not valid markup. This may 
cause problems for assistive technology. 

```
<div class="ons-collapsible__heading ons-js-collapsible-heading svelte-1gosn14" 
role="link" aria-controls="topic-0" aria-expanded="false" data-ga-action="Close 
panel">
[...] 
<button type="button" class="ons-btn ons-collapsible__btn ons-js-collapsible-
button ons-u-d-no@xxs@s ons-btn--secondary ons-btn--small" aria-controls="topic-0" 
data-ga-action="Close panel"> 
<span class="ons-btn__inner ons-js-collapsible-button-inner">Show</span> 
</button> 
[...]
</div> 
```

**Reply by jrbarnes9** [reply]
Raised #1996

**Comment by James-Nitron** [comment]
WCAG A - Incorrect use of ARIA

The ‘Hide this’ button that is found in the collapsible element contains an aria-hidden attribute. The aria-hidden attribute should not be used on elements that can receive keyboard-only focus. This may cause some disorientation to screen reader users that are able to ‘tab’ onto the element.

```
<button type="button" class="ons-btn ons-js-collapsible-button ons-btn--secondary ons-btn--small" aria-hidden="true" aria-controls="collapsible" data-ga-action="Open panel">
<span class="ons-btn__inner ons-js-collapsible-button-inner">Hide this</span>
<span class="ons-btn__context ons-u-vh">What is a photovoltaic system? content</span>
</button>
```
Suggestion remove the aria-hidden attribute.

**Reply by jrbarnes9** [reply]
Resolved by #2013

**Comment by LaurenDowling** [comment]
When CSS is turned off, the svg takes up a huge amount of space onscreen. This reduces the usability of the component. The same goes for the accordion component.
![image](https://user-images.githubusercontent.com/108019533/217626516-12bc2264-463f-4ff6-8489-e708edd96a0c.png)
Might be worth comparing with the gov one. The following is with css removed:
![image](https://user-images.githubusercontent.com/108019533/217626740-66008ca2-44f0-494d-bb31-aa358d88fc2f.png)

**Comment by Al-Davies** [comment]
**UR findings on collapsible**
tested as part of end-to-end Census testing Feb-Sept 2020

- These have been well tested via census EQ and census website. Are now widely used on the ONS - these reduce page length so page isn't over-daunting for users.
- Users like these - they don't have to read content they are not interested in, but understand how they work and can hide the content again should they wish.

**Comment by bothness** [comment]
Hi. Is there any update on the intention to revert this to be based on an HTML `<details>` component?

We have noted an additional issue in the current implementation that an attempt to find text within the page using CTRL-F does not reveal text within the current component, whereas it WILL find text within an HTML `<details>` component (automatically opening the twisty). You can see this desired behaviour in the [GOV UK details component](https://design-system.service.gov.uk/components/details/).

As we make increased use of these components on the new ONS website, it's important that users can find the information they need by this method. (Our user engagement thus far suggests that many users rely on CTRL-F to locate information on our pages.)

**Reply by rmccar** [reply]
Thanks for raising this. The reason why we aren't using the `<details>` implementation is that there was the need for details components that were closed on a page to automatically open and then close again when the page was printed with CTRL-P so that there was no content missed on the printed page. That is why we went down the JS route with a fallback for when JS isn't supported. We will raise a ticket to have a look into finding a solution to support this alongside the print functionality we already have.

**Reply by nimasmi** [reply]
You can achieve this without JS with the `::details-content` pseudo element.

```
@media print {
  details::details-content {
    display: contents;
  }
```

Demo: https://codepen.io/nimasmi/pen/gbbXjeG

It's pretty new, and Firefox doesn't support it yet, but it might be feasible soon. https://caniuse.com/?search=%3A%3Adetails-content

**Reply by rmccar** [reply]
Cheers @nimasmi, looks promising. Definitely something that we can look to implement soon. Would solve our issue perfectly

**Reply by bothness** [reply]
Hi @rmccar @nimasmi. Thanks for responding so quickly. It would be great if we can find a workable solution to this issue ahead of the rollout of the Wagtail CMS in July.
